compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -40,77 +40,137 @@ use tracing::{debug, trace};
 #[macro_use]
 mod pass_manager;
 
+use std::sync::LazyLock;
+
 use pass_manager::{self as pm, Lint, MirLint, MirPass, WithMinOptLevel};
 
-mod abort_unwinding_calls;
-mod add_call_guards;
-mod add_moves_for_packed_drops;
-mod add_retag;
-mod add_subtyping_projections;
-mod check_alignment;
-mod check_const_item_mutation;
-mod check_packed_ref;
-mod check_undefined_transmutes;
-// This pass is public to allow external drivers to perform MIR cleanup
-pub mod cleanup_post_borrowck;
-mod copy_prop;
-mod coroutine;
 mod cost_checker;
-mod coverage;
 mod cross_crate_inline;
-mod ctfe_limit;
-mod dataflow_const_prop;
-mod dead_store_elimination;
 mod deduce_param_attrs;
-mod deduplicate_blocks;
-mod deref_separator;
-mod dest_prop;
-pub mod dump_mir;
-mod early_otherwise_branch;
-mod elaborate_box_derefs;
-mod elaborate_drops;
 mod errors;
 mod ffi_unwind_calls;
-mod function_item_references;
-mod gvn;
-// Made public so that `mir_drops_elaborated_and_const_checked` can be overridden
-// by custom rustc drivers, running all the steps by themselves. See #114628.
-pub mod inline;
-mod instsimplify;
-mod jump_threading;
-mod known_panics_lint;
-mod large_enums;
 mod lint;
-mod lower_intrinsics;
-mod lower_slice_len;
-mod match_branches;
-mod mentioned_items;
-mod multiple_return_terminators;
-mod nrvo;
-mod post_drop_elaboration;
-mod prettify;
-mod promote_consts;
-mod ref_prop;
-mod remove_noop_landing_pads;
-mod remove_place_mention;
-mod remove_storage_markers;
-mod remove_uninit_drops;
-mod remove_unneeded_drops;
-mod remove_zsts;
-mod required_consts;
-mod reveal_all;
-mod sanity_check;
 mod shim;
 mod ssa;
-// This pass is public to allow external drivers to perform MIR cleanup
-pub mod simplify;
-mod simplify_branches;
-mod simplify_comparison_integral;
-mod single_use_consts;
-mod sroa;
-mod unreachable_enum_branching;
-mod unreachable_prop;
-mod validate;
+
+macro_rules! declare_passes {
+    (
+        $(
+            $vis:vis mod $mod_name:ident : $($pass_name:ident $( { $($ident:ident),* } )?),+;
+        )*
+    ) => {
+        $(
+            $vis mod $mod_name;
+            $(
+                // Make sure the type name is correct
+                #[allow(unused_imports)]
+                use $mod_name::$pass_name as _;
+            )+
+        )*
+
+        #[cfg(debug_assertions)]
+        static PASS_NAMES: LazyLock<Vec<String>> = LazyLock::new(|| vec![
+            // Fake marker pass
+            "PreCodegen".to_string(),
+            $(
+                $(
+                    stringify!($pass_name).to_string(),
+                    $(
+                        $(
+                            $mod_name::$pass_name::$ident.name().to_string(),
+                        )*
+                    )?
+                )+
+            )*
+        ]);
+    };
+}
+
+declare_passes! {
```

Comment:
Suggestion: could you add some brief comments on this to explain what the macro *intends* to do? I.e. by example:

- `mod abort_unwinding_calls : AbortUnwindingCalls;`: pass inside own module, one single pass named `AbortUnwindingCalls`
- `mod add_call_guards : AddCallGuards { AllCallEdges, CriticalCallEdges };`: one pass group named `AddCallGuards` which contains two constituent passes named `AllCallEdges` and `CriticalCallEdges`.
- `mod simplify : SimplifyCfg { Initial, /* omitted */ }, SimplifyLocals { BeforeConstProp, /* omitted */ }`: one module, two pass groups `SimplifyCfg` and `SimplifyLocals`, each with their own constituent passes.

and also why we need this macro (so we can know statistically all of the passes).